File: app/src/test/java/com/github/fo2rist/mclaren/repository/McLarenFeedRepositoryImplTest.java
package com.github.fo2rist.mclaren.repository;

import com.github.fo2rist.mclaren.repository.FeedRepositoryEventBus.LoadingEvent;
import com.github.fo2rist.mclaren.web.FeedHistoryPredictor;
import com.github.fo2rist.mclaren.web.FeedWebService.FeedRequestCallback;
import com.github.fo2rist.mclaren.web.McLarenFeedWebService;
import java.net.URL;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static com.github.fo2rist.mclaren.testdata.McLarenFeedResponse.REAL_FEED_RESPONSE;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = 21)
public class McLarenFeedRepositoryImplTest {
    //TODO remove tests duplications with StoryStream Repo test. 2018.11.25

    private McLarenFeedRepositoryImpl repository;
    private McLarenFeedWebService mockWebService;
    private FeedRepositoryEventBus mockEventBus;
    private FeedHistoryPredictor mockHistoryPredictor;

    @Before
    public void setUp() {
        mockWebService = mock(McLarenFeedWebService.class);
        mockEventBus = mock(FeedRepositoryEventBus.class);
        mockHistoryPredictor = mock(FeedHistoryPredictor.class);

        repository = new McLarenFeedRepositoryImpl(mockWebService, mockEventBus, mockHistoryPredictor);
    }

    @Test
    public void test_loadLatest_startLoading_and_firesLoadStartEvent() {
        repository.loadLatest();

        verify(mockEventBus).publish(any(LoadingEvent.LoadingStarted.class));
        verify(mockWebService).requestLatestFeed(any(FeedRequestCallback.class));
    }

    @Test
    public void testHistoryLoadingCallPredictorFirst() {
        when(mockHistoryPredictor.isFirstHistoryPageKnown()).thenReturn(false);
        when(mockHistoryPredictor.getFirstHistoryPage()).thenReturn(-1);

        repository.loadNextHistory();

        verify(mockHistoryPredictor).startPrediction();
        verify(mockWebService, never()).requestFeedPage(anyInt(), any(FeedRequestCallback.class));
    }

    @Test
    public void testHistoryLoadingCallPredictedPage() {
        when(mockHistoryPredictor.isFirstHistoryPageKnown()).thenReturn(true);
        when(mockHistoryPredictor.getFirstHistoryPage()).thenReturn(1000);

        repository.loadNextHistory();

        verify(mockHistoryPredictor, never()).startPrediction();
        verify(mockHistoryPredictor).getFirstHistoryPage();
        verify(mockWebService).requestFeedPage(anyInt(), any(FeedRequestCallback.class));
    }

    @Test
    public void test_PrepareForHistoryLoading_calledWhenHistoryIsUnknown() {
        when(mockHistoryPredictor.isFirstHistoryPageKnown()).thenReturn(false);

        repository.prepareForHistoryLoading();

        verify(mockHistoryPredictor).startPrediction();
    }

    @Test
    public void test_PrepareForHistoryLoading_notCalledWhenHistoryIsKnown() {
        when(mockHistoryPredictor.isFirstHistoryPageKnown()).thenReturn(true);

        repository.prepareForHistoryLoading();

        verify(mockHistoryPredictor, never()).startPrediction();
    }

    @Test
    public void test_onSuccess_firesLoadFinishEvents() {
        doAnswer(new Answer() {
            @Override
            public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
                FeedRequestCallback callback = invocationOnMock.getArgument(0);
                callback.onSuccess(new URL("http://empty.url"), 0, 200, REAL_FEED_RESPONSE);
                return null;
            }
        }).when(mockWebService).requestLatestFeed(any(FeedRequestCallback.class));

        repository.loadLatest();

        verify(mockEventBus).publish(any(LoadingEvent.LoadingFinished.class));
        verify(mockEventBus).publish(any(LoadingEvent.FeedUpdateReady.class));
    }

    @Test
    public void test_onFailure_firesLoadFinishWithErrorEvents() {
        doAnswer(new Answer() {
            @Override
            public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
                FeedRequestCallback callback = invocationOnMock.getArgument(0);
                callback.onFailure(new URL("http://empty.url"), 0, 400, null);
                return null;
            }
        }).when(mockWebService).requestLatestFeed(any(FeedRequestCallback.class));

        repository.loadLatest();

        verify(mockEventBus).publish(any(LoadingEvent.LoadingFinished.class));
        verify(mockEventBus).publish(any(LoadingEvent.LoadingError.class));
    }
}

File: app/src/test/java/com/github/fo2rist/mclaren/repository/McLarenFeedRepositoryImplTest.kt
package com.github.fo2rist.mclaren.repository

import com.github.fo2rist.mclaren.repository.FeedRepositoryEventBus.LoadingEvent
import com.github.fo2rist.mclaren.testdata.McLarenFeedResponse.REAL_FEED_RESPONSE
import com.github.fo2rist.mclaren.web.FeedHistoryPredictor
import com.github.fo2rist.mclaren.web.McLarenFeedWebService
import com.github.fo2rist.mclaren.web.utils.BadResponse
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.whenever
import kotlinx.coroutines.runBlocking
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.Mock
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import java.net.URL

//TODO remove tests duplications with StoryStream Repo test. 2018.11.25
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [21])
class McLarenFeedRepositoryImplTest {

    @Mock
    private lateinit var mockWebService: McLarenFeedWebService
    @Mock
    private lateinit var mockEventBus: FeedRepositoryEventBus
    @Mock
    private lateinit var mockHistoryPredictor: FeedHistoryPredictor

    private lateinit var repository: McLarenFeedRepositoryImpl

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)

        repository = McLarenFeedRepositoryImpl(mockWebService, mockEventBus, mockHistoryPredictor)
    }

    @Test
    fun test_loadLatest_startLoading_and_firesLoadStartEvent() = runBlocking<Unit> {
        repository.loadLatestPage()

        verify(mockEventBus).publish(any<LoadingEvent.LoadingStarted>())
        verify(mockWebService).requestLatestFeed()
    }

    @Test
    fun testHistoryLoadingCallPredictorFirst() {
        whenever(mockHistoryPredictor.isFirstHistoryPageKnown).thenReturn(false)
        whenever(mockHistoryPredictor.firstHistoryPage).thenReturn(-1)

        repository.loadNextPage()

        verify(mockHistoryPredictor).startPrediction()
        verify(mockWebService, never()).requestFeedPage(anyInt(), any())
    }

    @Test
    fun testHistoryLoadingCallPredictedPage() {
        whenever(mockHistoryPredictor.isFirstHistoryPageKnown).thenReturn(true)
        whenever(mockHistoryPredictor.firstHistoryPage).thenReturn(1000)

        repository.loadNextPage()

        verify(mockHistoryPredictor, never()).startPrediction()
        verify(mockHistoryPredictor).firstHistoryPage
        verify(mockWebService).requestFeedPage(anyInt(), any())
    }

    @Test
    fun test_PrepareForHistoryLoading_calledWhenHistoryIsUnknown() {
        whenever(mockHistoryPredictor.isFirstHistoryPageKnown).thenReturn(false)

        repository.prepareForHistoryLoading()

        verify(mockHistoryPredictor).startPrediction()
    }

    @Test
    fun test_PrepareForHistoryLoading_notCalledWhenHistoryIsKnown() {
        whenever(mockHistoryPredictor.isFirstHistoryPageKnown).thenReturn(true)

        repository.prepareForHistoryLoading()

        verify(mockHistoryPredictor, never()).startPrediction()
    }

    @Test
    fun test_onSuccess_firesLoadFinishEvents() = runBlocking {
        whenever(mockWebService.requestLatestFeed()).thenReturn(REAL_FEED_RESPONSE)

        repository.loadLatestPage()

        verify(mockEventBus).publish(any<LoadingEvent.LoadingStarted>())
        verify(mockEventBus).publish(any<LoadingEvent.FeedUpdateReady>())
        verify(mockEventBus).publish(any<LoadingEvent.LoadingFinished>())
    }

    @Test
    fun test_onFailure_firesLoadFinishWithErrorEvents() = runBlocking {
        whenever(mockWebService.requestLatestFeed())
                .thenThrow(BadResponse(URL("http://empty.url"), 400))

        repository.loadLatestPage()

        verify(mockEventBus).publish(any<LoadingEvent.LoadingStarted>())
        verify(mockEventBus).publish(any<LoadingEvent.LoadingError>())
        verify(mockEventBus).publish(any<LoadingEvent.LoadingFinished>())
    }
}
